### PR TITLE
Provide constructors for SkeletalFinSets

### LINKED ISSUE #30: Use SkeletalFinSets as a workhorse
I would like to use `SkeletalFinSets` as the range of a homomorphism structure of categories that only have finitely many morphisms between two objects (like finite groups or monoids as categories, or acyclic quivers as categories). For this, some constructions of `SkeletalFinSets` need to be as fast as possible (in particular the morphism constructor).

Is it possible to provide a constructor for `SkeletalFinSets` for the user, so that it is possible to pipe options to the constructor like the one introduced in #405 that immediately give a vast speed-up
in my use case?

### PR DESCRIPTION
**Pull request by sebastianpos**
This closes #30.

Moreover, it includes #31, so #31 should be merged first.

### CODE REVIEW

**Comment by sebastianpos** [comment]
Sorry, I just realized that for every new variant of skeletal finite sets, no methods will be added.

I'm working on a fix.

**Comment by sebastianpos** [comment]
I hopefully fixed my mistake.

File: gap/SkeletalFinSetsForCAP.gd [hunk 1]
```diff
@@ -42,6 +52,19 @@ DeclareAttribute( "AsList",
 
 #! @Section Skeletal Constructors
 
+#! @Description
+#!  Construct a category of skeletal finite sets.
+#! @Returns a &CAP; category
+DeclareOperation( "CategoryOfSkeletalFinSets", [] );
```
Comment (zickgraf) [inline gap/SkeletalFinSetsForCAP.gd]: Should this be an operation or simply a global function?

**Comment by sebastianpos** [comment]
I provided new commits concerning your general points 1 and 2.

Concerning Point 2:
If I replace the line
`BindGlobal( "SkeletalFinSets", CategoryOfSkeletalFinSets() );`
by the two lines
```
DeclareGlobalVariable( "SkeletalFinSets" );

InstallValue( SkeletalFinSets, CategoryOfSkeletalFinSets() );
```
in order to provide documentation, the tests yield an error that I don't understand. Maybe you can reproduce this error and make sense out of it.

**Comment by sebastianpos** [comment]
Concerning point 3:

It depends on your application. If you want to create lots of fin sets, then caching might be a bad idea. In my applications, I don't need many fin sets, and so it's good to have fixed objects for each number, since this makes e.g. comparisons quicker.

I leave the decision to you, since it is your package.

File: gap/SkeletalFinSetsForCAP.gd [hunk 2]
```diff
@@ -42,6 +52,19 @@ DeclareAttribute( "AsList",
...
+
+#! @Description
+#!  Construct a skeletal finite set residing in
+#!  the given category of skeletal finite sets <A>C</A>
+#!  of order given be the nonnegative integer <A>n</A>.
+#! @Arguments C, n
+#! @Returns a &CAP; object
+KeyDependentOperation( "FinSet", IsCategoryOfSkeletalFinSets, IsInt, ReturnTrue );
```
[the remark this answers is not in the document]
Comment (sebastianpos) [reply gap/SkeletalFinSetsForCAP.gd]: By now, your philosophy seemed to be that you let any input pass through the `FinSet` constructor, and use the `Assert` machinery in order test for well-definedness.

Again, it's your decision as a package author whether you want to change that. I would probably leave it as it is.

File: gap/SkeletalFinSetsForCAP.gi [hunk 3]
```diff
@@ -4,25 +4,43 @@
...
-
-AddMorphismRepresentation( SkeletalFinSets, IsSkeletalFiniteSetMap );
+##
+InstallMethod( CategoryOfSkeletalFinSets,
+               [ ],
+               
+  function( )
+    local overhead_option, cat;
+    
+    overhead_option := CAP_INTERNAL_RETURN_OPTION_OR_DEFAULT( "overhead", true );
```
[the remark this answers is not in the document]
Comment (sebastianpos) [reply gap/SkeletalFinSetsForCAP.gi]: Pushed a commit for `FinalizeCategory`.

So far I don't know other options. Of course there can be lots of such options in the future.

File: gap/SkeletalFinSetsForCAP.gd (hunk 1 above)
Comment (sebastianpos) [reply gap/SkeletalFinSetsForCAP.gd]: I don't know if there might be versions of higher arity of this operation in the future.

You decide.

File: gap/SkeletalFinSetsForCAP.gd (hunk 2 above)
[the remark this answers is not in the document]
Comment (zickgraf) [reply gap/SkeletalFinSetsForCAP.gd]: Okay, then we leave it, for the last nanosecond of performance :D

File: gap/SkeletalFinSetsForCAP.gi (hunk 3 above)
[the remark this answers is not in the document]
Comment (zickgraf) [reply gap/SkeletalFinSetsForCAP.gi]: Thanks :-)

File: gap/SkeletalFinSetsForCAP.gd (hunk 1 above)
Comment (zickgraf) [reply gap/SkeletalFinSetsForCAP.gd]: I have talked to Max and he has no concerns about a 0-ary operation, so we can leave it as it is.

**Comment by zickgraf** [comment]
Concerning point 2: This is unexpected. I will look into this after this PR is merged.

Concerning point 3: sorry, I should have phrased this more precisely, the question was meant with regard to your concrete application. Since it helps your applications and I'm not aware of other relevant applications, it seems a good idea to use `KeyDependenOperation`.

**Comment by mohamed-barakat** [comment]
@zickgraf: Is it possible to test the performance change with Jenkins and post the links here?

**Pull request merged**